### ISSUE TITLE
LRC

### ISSUE THREAD

**Issue by juancarloslin** [Psychopoulet/split-frames#15]
We have a need to split-frames but also check LRC
So STX........ETX LRC 
LRC calculated without STX.
Will be nice to allow checkLRC option in the start with / end with so take the ETX followed  byte to validate LRC, then emmit data.

**Comment by Psychopoulet** [comment]
the point is, I cannot add a specific comportment on it because it will no longer be generic. I work personaly with a lot of protocols, and each one works with its own way...
for your purpose, you need to check LRC, with ETX and without STX (Concert protocol, or something like this ?), but what about protocols with STX, or without ETX, or LRC before ETX (or after STX) ?
and what about protocols which use MSB for double checking ? I think it may be more valuable to inherit this package to create a specific parser for a specific comportment...

I was thinking about that (to add one or more bit after the and, for example, which can be parsed by inheritence), but I don't have enougth time at this moment for that... there is so much protocols ! so packages to create ! ^^

**Comment by juancarloslin** [comment]
Hi, 

Thanks for your replay, 
I was thinking about a Generic LRC, as ANY, so then user can check on his own code. 
ANY means any value, but, a byte., can be represented as 0x00 


If LRC is calculated at the end of ETX as our case should be 

Start with STX 
End with ETX, ANY 

so return STX,.......,ETX,LRC 

if LRC calculated at the start of STX and include ETX then 

Start with ANY, STX 
End with ETX 

so return LRC,STX,.......,ETX 

if LRC is 2 bytes, calculated at the start of STX and include ETX then 

Start with ANY, ANY, STX 
End with ETX 

so return LRC1, LRC2,STX,.......,ETX 

this may works. 


The ANY switch will be very powerfull for procesing frames. 


Juan Carlos Lin 
Unisoft SA 
+595-993-288330 


De: "Sébastien Vidal" <notifications@github.com> 
Para: "Psychopoulet/split-frames" <split-frames@noreply.github.com> 
CC: "Juan Carlos Lin" <lin@unisoft.com.py>, "Author" <author@noreply.github.com> 
Enviados: Domingo, 26 de Agosto 2018 16:44:20 
Asunto: Re: [Psychopoulet/split-frames] LRC (#15) 



the point is, I cannot add a specific comportment on it because it will no longer be generic. I work personaly with a lot of protocols, and each one works with its own way... 
for your purpose, you need to check LRC, with ETX and without STX (Concert protocol, or something like this ?), but what about protocols with STX, or without ETX, or LRC before ETX (or after STX) ? 
and what about protocols which use MSB for double checking ? I think it may be more valuable to inherit this package to create a specific parser for a specific comportment... 

I was thinking about that (to add one or more bit after the and, for example, which can be parsed by inheritence), but I don't have enougth time at this moment for that... there is so much protocols ! so packages to create ! ^^

**Comment by Psychopoulet** [comment]
as I said, you have so many protocols ! there is no way to know how to compute it in a generic way.

for example :

STX < computed> ETX LRC
STX < computed ETX > LRC
< STX computed ETX > LRC
< STX computed > ETX LRC

with Last Significant Byte & Most Significant Byte :
STX < computed > ETX LSB MSB
STX < computed > ETX MSB LSB
STX < computed > LSB MSB ETX
STX < computed LSB MSB > ETX

etc... etc... etc...

I can extract some specific bytes (STX +x, ETX +x, ETX -x, etc...), but to know what it means and how to compute it, you have to inheritate it

**Comment by juancarloslin** [comment]
Yes, 

the idias is allow to extract the byte, then inside the code that consume the split-fram, check if is valid. 

so the ANY siwtch will indicate that that byte sould be come withe the chunck frame. 
so if 


startwith STX 
endwith ETX ANY 

will come with STX....EXT + 1 byte with any value. 

if startwith ANY ANY STX 
edwtih ETX 

will come with 2 bytes with any value + STX...ETX 

then the consumer of split-frame can check is the package is valid or no. 

or you can call it LRC, just indicating LRC is no validated at split-frame level. 


Juan Carlos Lin 
Unisoft SA 
+595-993-288330 


De: "Sébastien Vidal" <notifications@github.com> 
Para: "Psychopoulet" <split-frames@noreply.github.com> 
CC: "Juan Carlos Lin" <lin@unisoft.com.py>, "Author" <author@noreply.github.com> 
Enviados: Martes, 28 de Agosto 2018 15:20:43 
Asunto: Re: [Psychopoulet/split-frames] LRC (#15) 



as I said, you have so many protocols ! there is no way to know how to compute it in a generic way. 

for example : 

STX < computed> ETX LRC 
STX < computed ETX > LRC 
< STX computed ETX > LRC 
< STX computed > ETX LRC 

with Last Significant Byte & Most Significant Byte : 
STX < computed > ETX LSB MSB 
STX < computed > ETX MSB LSB 
STX < computed > LSB MSB ETX 
STX < computed LSB MSB > ETX 

etc... etc... etc... 

I can extract some specific bytes (STX +x, ETX +x, ETX -x, etc...), but to know what it means and how to compute it, you have to inheritate it

**Comment by Psychopoulet** [comment]
Yep. as soon as it's possible, I'll do that (just had a baby ^^). Probably next week.

**Comment by Psychopoulet** [comment]
it's my second one, so... we all sleep all free time ^^

**Comment by Psychopoulet** [comment]
I started to rewrite it. it's a new main version with logic change : it will give you the splited frame as it is, with all its bits (STX <data> ETX <vontrol bit>) and it will no remove escaped bits.

If you want parse escaped bits, remove start/end bits or compute control bits (LRC, LSB, MSB, etc...), you will have to pipe it and compute it yourself in your class.

- advantage: it will be more coherent, only split frame without computing, and you will be able to compute it as you want.
- disadvantage: you will have to extract escaped bits and remove start/end bits by yourself

**Comment by Psychopoulet** [comment]
done.

**Comment by juancarloslin** [comment]
Good idea, 

I also checked, the example code need to be fix 


function computeLRC ( frame ) { 

should be 

function _ computeLRC ( frame ) { 


Juan Carlos Lin 
Unisoft SA 
+595-993-288330 


De: "Sébastien Vidal" <notifications@github.com> 
Para: "Psychopoulet" <split-frames@noreply.github.com> 
CC: "Juan Carlos Lin" <lin@unisoft.com.py>, "Author" <author@noreply.github.com> 
Enviados: Martes, 4 de Septiembre 2018 21:38:16 
Asunto: Re: [Psychopoulet/split-frames] LRC (#15) 



I started to rewrite it. it's a new main version with logic change : it will give you the splited frame as it is, with all its bits (STX ETX ) and it will no remove escaped bits. 

If you want parse escaped bits, remove start/end bits or compute control bits (LRC, LSB, MSB, etc...), you will have to pipe it and compute it yourself in your class. 

    * advantage: it will be more coherent, only split frame without computing, and you will be able to compute it as you want. 
    * disadvantage: you will have to extract escaped bits and remove start/end bits by yourself
